In [1]:
import numpy as num
import pandas as pd
import os
import nltk as np
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
import re
from nltk.tokenize import word_tokenize 
from nltk.stem.porter import PorterStemmer
from nltk.stem import WordNetLemmatizer
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle

#Libraries that check the accuracy of the model over test set
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score

Code to Determine the Proper directory to your data

In [2]:
for dirname, _, filenames in os.walk('./'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

./FN_Model.h5
./Model.ipynb
./.git\COMMIT_EDITMSG
./.git\config
./.git\description
./.git\HEAD
./.git\index
./.git\hooks\applypatch-msg.sample
./.git\hooks\commit-msg.sample
./.git\hooks\fsmonitor-watchman.sample
./.git\hooks\post-update.sample
./.git\hooks\pre-applypatch.sample
./.git\hooks\pre-commit.sample
./.git\hooks\pre-merge-commit.sample
./.git\hooks\pre-push.sample
./.git\hooks\pre-rebase.sample
./.git\hooks\pre-receive.sample
./.git\hooks\prepare-commit-msg.sample
./.git\hooks\push-to-checkout.sample
./.git\hooks\update.sample
./.git\info\exclude
./.git\logs\HEAD
./.git\logs\refs\heads\main
./.git\logs\refs\remotes\origin\main
./.git\objects\03\805733ab6dea2d4666a05bacc81d40fc6e639d
./.git\objects\0a\b53fd67cb79705c847fa93d44f501a9d36adb5
./.git\objects\16\e2f04fa018fa8f027d6fc3afde6d6e89f7df6e
./.git\objects\1d\905edc495e788a78623ff1f1f29c94eb7a6494
./.git\objects\20\167fddd811aa4f7b9997c5deb494edd5e9771d
./.git\objects\45\128f7b518f6dc52a7c576d2f1f5ac3cb0db99c
./.git\object

Opening Data using Pandas

In [3]:
Fake_data = pd.read_csv('./Fake/Fake.csv')
True_data = pd.read_csv('./True/True.csv')

Data Clean-up....Begining

In [4]:
Fake_data.head(2) #THE initial form in which our data is in.

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   

                date  
0  December 31, 2017  
1  December 31, 2017

In [5]:
Fake_data =Fake_data.drop(columns=["date","subject","title"]) #Removing some unwanted columns.

In [6]:
True_data = True_data.drop(columns=["date","subject","title"])

Assigning Tags for easy differetiation of News types (either Fake or True) True news = 1 and Fake news = 0
(i.e. Any value approaching 1 indicates Fake news and any Value approaching 0 indicates True news)

In [7]:
True_data['fake'] = 1
Fake_data['fake'] = 0

In [8]:
Fake_data.head(2) #Previewing our new data.

text  fake
0  Donald Trump just couldn t wish all Americans ...     0
1  House Intelligence Committee Chairman Devin Nu...     0

In [9]:
True_data.head()

text  fake
0  WASHINGTON (Reuters) - The head of a conservat...     1
1  WASHINGTON (Reuters) - Transgender people will...     1
2  WASHINGTON (Reuters) - The special counsel inv...     1
3  WASHINGTON (Reuters) - Trump campaign adviser ...     1
4  SEATTLE/WASHINGTON (Reuters) - President Donal...     1

In [10]:
Fake_data.shape

(23481, 2)

In [11]:
True_data.shape

(21417, 2)

In [12]:
word = "My name is David Adlai Nettey. I do Computer Science"

In [13]:
token = np.word_tokenize(word)

In [14]:
print(token)

['My', 'name', 'is', 'David', 'Adlai', 'Nettey', '.', 'I', 'do', 'Computer', 'Science']


In [15]:
trial = Fake_data['text'][1]

In [16]:
trial = trial.lower() # Converting from uppercase to lowercase


In [17]:
print(trial)

house intelligence committee chairman devin nunes is going to have a bad day. he s been under the assumption, like many of us, that the christopher steele-dossier was what prompted the russia investigation so he s been lashing out at the department of justice and the fbi in order to protect trump. as it happens, the dossier is not what started the investigation, according to documents obtained by the new york times.former trump campaign adviser george papadopoulos was drunk in a wine bar when he revealed knowledge of russian opposition research on hillary clinton.on top of that, papadopoulos wasn t just a covfefe boy for trump, as his administration has alleged. he had a much larger role, but none so damning as being a drunken fool in a wine bar. coffee boys  don t help to arrange a new york meeting between trump and president abdel fattah el-sisi of egypt two months before the election. it was known before that the former aide set up meetings with world leaders for trump, but team tru

Function necessary for cleaning our Data

In [18]:
def remove_punctuation(text):  #Removing any kind of punctuation present in the data
    return re.sub(r'[^\w\s]','', str(text))

def remove_urls(text):  #Removing urls from the data
    return re.sub(r"http\S+", "", str(text))
                  
def remove_stopwords(text):  #Removing stopwords(eg. this, that, am, be etc)
    stop = stopwords.words("english")
    final_text = []
    for i in str(text).split():
        if i.strip() not in stop:
            final_text.append(i.strip())
    return " ".join(final_text)
                  
def tokenize(text):  
    tokens = re.split('\W+',text) #W+ means that either a word character (A-Z) or a dash(-) can go there.
    return tokens

def tokenize_words(text):  #Converting all the text to lower case
    return word_tokenize(text.lower())

                                  
def stemming(text):  #Converting the words into their stem form
    porter_stemmer = PorterStemmer()
    return porter_stemmer.stem(str(text))
                  
def lemmatization(text):  #Applying Lemaatization i.e., converting words into their lemma
    wordnet_lemmatizer = WordNetLemmatizer()
    return wordnet_lemmatizer.lemmatize(str(text))

In [19]:
#Function that executes our data cleaning...
def clean_text(text):
    text = remove_stopwords(text)
    text = remove_punctuation(text)
    text = remove_urls(text)
    text = tokenize_words(text)
    text = stemming(text)
    return lemmatization(text)

In [20]:
True_data['text'] = True_data['text'].apply(lambda x: clean_text(x))

In [21]:
Fake_data['text'] = Fake_data['text'].apply(lambda x: clean_text(x))

In [22]:
Fake_data['text'][0]

"['donald', 'trump', 'wish', 'americans', 'happy', 'new', 'year', 'leave', 'that', 'instead', 'give', 'shout', 'enemies', 'haters', 'dishonest', 'fake', 'news', 'media', 'the', 'former', 'reality', 'show', 'star', 'one', 'job', 'it', 'as', 'country', 'rapidly', 'grows', 'stronger', 'smarter', 'i', 'want', 'wish', 'friends', 'supporters', 'enemies', 'haters', 'even', 'dishonest', 'fake', 'news', 'media', 'happy', 'healthy', 'new', 'year', 'president', 'angry', 'pants', 'tweeted', '2018', 'great', 'year', 'america', 'as', 'country', 'rapidly', 'grows', 'stronger', 'smarter', 'i', 'want', 'wish', 'friends', 'supporters', 'enemies', 'haters', 'even', 'dishonest', 'fake', 'news', 'media', 'happy', 'healthy', 'new', 'year', '2018', 'great', 'year', 'america', 'donald', 'j', 'trump', 'realdonaldtrump', 'december', '31', '2017trump', 'tweet', 'went', 'welll', 'expectwhat', 'kind', 'president', 'sends', 'new', 'year', 'greeting', 'like', 'despicable', 'petty', 'infantile', 'gibberish', 'only', 

Combining both fake and true news to form one Dataset

In [23]:
News_data = pd.concat([True_data,Fake_data])

In [24]:
News_data.shape

(44898, 2)

In [25]:
News_data.tail(2)

text  fake
23479  ['21st', 'century', 'wire', 'says', 'al', 'jaz...     0
23480  ['21st', 'century', 'wire', 'says', 'as', '21w...     0

Shuffling our dataset

In [26]:
News_data = shuffle(News_data,random_state=5)

Data Clean-up...Completed

Building RNN Model

In [27]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.callbacks import EarlyStopping

RNN Model

In [28]:
x_train,x_test,y_train,y_test =  train_test_split(News_data['text'],News_data['fake'],test_size=0.20,random_state=8)

In [29]:
x_train.shape

(35918,)

In [30]:
#Tokenize the words into vectors because we can only give numerical data as input to the model
max_vocab = 5000
tokenizer = Tokenizer(num_words = max_vocab)
tokenizer.fit_on_texts(x_train)

x_train = tokenizer.texts_to_sequences(x_train)
x_test = tokenizer.texts_to_sequences(x_test)

#Padding is applied so that we get the same length of input for each article

x_train = pad_sequences(x_train, padding = "post", maxlen = 25)
x_test = pad_sequences(x_test, padding = "post", maxlen = 25)

In [31]:
x_train.shape

(35918, 25)

In [32]:
embedding_vector_features = 256
model=tf.keras.Sequential()
model.add(tf.keras.layers.Embedding(5000, embedding_vector_features, input_length=25))
model.add(tf.keras.layers.SpatialDropout1D(0.2))
model.add(tf.keras.layers.LSTM(64, dropout=0.2,recurrent_dropout=0.2))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(100, activation = 'relu'))
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adagrad', metrics=['accuracy'])

In [33]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 25, 256)           1280000   
                                                                 
 spatial_dropout1d (SpatialD  (None, 25, 256)          0         
 ropout1D)                                                       
                                                                 
 lstm (LSTM)                 (None, 64)                82176     
                                                                 
 dropout (Dropout)           (None, 64)                0         
                                                                 
 dense (Dense)               (None, 100)               6500      
                                                                 
 dense_1 (Dense)             (None, 1)                 101       
                                                        

In [34]:
model.fit(x_train, y_train, epochs = 6, validation_split = 0.10, batch_size = 30, shuffle = True, verbose=1)

Epoch 1/6
1078/1078 [==============================] - 61s 52ms/step - loss: 0.6888 - accuracy: 0.5490 - val_loss: 0.6854 - val_accuracy: 0.5354
Epoch 2/6
1078/1078 [==============================] - 54s 50ms/step - loss: 0.6830 - accuracy: 0.5579 - val_loss: 0.6789 - val_accuracy: 0.5649
Epoch 3/6
1078/1078 [==============================] - 53s 49ms/step - loss: 0.6755 - accuracy: 0.5963 - val_loss: 0.6693 - val_accuracy: 0.6269
Epoch 4/6
1078/1078 [==============================] - 54s 50ms/step - loss: 0.6637 - accuracy: 0.6750 - val_loss: 0.6539 - val_accuracy: 0.7266
Epoch 5/6
1078/1078 [==============================] - 55s 51ms/step - loss: 0.6447 - accuracy: 0.7579 - val_loss: 0.6283 - val_accuracy: 0.8238
Epoch 6/6
1078/1078 [==============================] - 54s 50ms/step - loss: 0.6120 - accuracy: 0.8164 - val_loss: 0.5842 - val_accuracy: 0.8402


In [35]:
model.evaluate(x_test,y_test)

281/281 [==============================] - 2s 8ms/step - loss: 0.5851 - accuracy: 0.8404


[0.5851209759712219, 0.8404231667518616]

Model Accuracy after testing on unseen data is 84.04%

In [36]:
predict = model.predict(x_test)

With our Model Prediction, 1 == Fake news and 0 == Factual news. eg: if the model predicts 80% it definately means Fake news and if the model predicts 20% its definately Factual news.

In [37]:
print(predict)

[[0.5352786 ]
 [0.52618843]
 [0.5328309 ]
 ...
 [0.3464765 ]
 [0.5159141 ]
 [0.5157778 ]]


In [38]:
def news_preperation(news:str):
    news = clean_text(news)
    tokenizer.fit_on_texts(news)
    news = tokenizer.texts_to_sequences(news)
    news = pad_sequences(news, padding= "post", maxlen = 25)
    results = model.predict(news)
    result = results[0]
    return "{:.2f}".format(result[0]*100)

In [40]:
res = news_preperation("There is a War in Ukraine")
res = float(res)
true = res
false = 100.00 - res
false = "{:.2f}".format(false)
print("Fake news Probability: "+ str(false) + " True News Probability: "+ str(true))

Fake news Probability: 82.62 True News Probability: 17.38


Saving the AI Model.

In [41]:
model.save("Model.h5")

Loading our Saved Model.

In [46]:
Fmodel = tf.keras.models.load_model("Model.h5")

In [48]:
Fmodel.evaluate(x_test,y_test)

281/281 [==============================] - 2s 9ms/step - loss: 0.5851 - accuracy: 0.8404


[0.5851209759712219, 0.8404231667518616]